### PR TITLE
[#37] Send email notifications to unselected pickers

### LINKED ISSUE #37: Send automatic emails for non-selected pickers
From: https://github.com/tiagovaz/saskatoon/issues/179

### PR DESCRIPTION
**Pull request by victorphoenix3**
Fixes #37 

Signed-off-by: Jayati Shrivastava <gaurijove@gmail.com>

### CODE REVIEW

**Comment by victorphoenix3** [comment]
Tested PR using console backend for django mails.

File: saskatoon/harvest/signals.py [hunk 1]
```diff
@@ -37,6 +38,30 @@ def changed_by(sender, instance, **kwargs):
     else:
         instance.changed_by = None
 
+def notify_unselected_pickers(sender, instance, **kwargs):
+    if instance.id:
+        original_instance = sender.objects.get(id=instance.id)
+        if instance.status == "Ready" and original_instance.status != "Ready":
+            unselected_pickers = instance.get_unselected_pickers()
```
Comment (tristanlatr) [inline saskatoon/harvest/signals.py:45]: So people that cancelled their participation will still receive an email ?

File: saskatoon/harvest/signals.py [hunk 2]
```diff
@@ -37,6 +38,30 @@ def changed_by(sender, instance, **kwargs):
...
+            email_list = list()
+            mail_subject = _("[Saskatoon] Not Selected as a picker for upcoming Harvest")
+            for p in unselected_pickers:
```
Comment (tristanlatr) [inline saskatoon/harvest/signals.py:48]: Looks like all unselected pickers will be notified each time the pick leaders edits the harvest ?
Maybe we could check and send the emails only when the harvest it passing from the status "Date scheduled" to "Ready". At this time, all required pickers should have been accepted.

**Review by tristanlatr — COMMENTED** [review]
Looks good, I have a couple of questions none the less.

Thanks a lot

File: saskatoon/harvest/signals.py (hunk 1 above)
Comment (victorphoenix3) [reply saskatoon/harvest/signals.py:45]: No, pickers that cancelled will not be notified. Only pickers that got rejected or those which are still pending will receive the emails.

File: saskatoon/harvest/signals.py [hunk 3]
```diff
@@ -37,6 +38,30 @@ def changed_by(sender, instance, **kwargs):
...
+                picker_name = p.picker.name()
+                picker_email = list()
+                picker_email.append(p.picker.email())
+                harvest_details = instance.__str__()
+                message = (_("Hi") + " " + picker_name + ",\n\n" +
+                    _("You have not been selected by the pick leader " +
+                    "to participate in the ") + harvest_details + ". " +
+                    _("You may still be contacted by the pick-leader " +
+                    "if other pickers cancel.") + "\n\n" +
```
Comment (tristanlatr) [inline saskatoon/harvest/signals.py:57]: I would add a note like the following: Your email address is associated to a volunteer profile, re-using the same email next time you request to participate in a pick will prioritize you over pickers that have been participating more over the season.

But this is a suggestion, and @2ynn might want to add 2 cents to that.

In practice, this is decided by the pick leader, and they might choose anyone, but there is formation about this in the guides.

File: saskatoon/harvest/signals.py [hunk 4]
```diff
@@ -38,6 +38,32 @@ def changed_by(sender, instance, **kwargs):
...
+                    _("We are sorry but enough pickers have already been " +
+                    "selected for the ") + harvest_details + ". " +
+                    _("You may still be contacted by the pick-leader " +
+                    "if some of them end up cancelling. We will do our best to " +
+                    "prioritize your participation next time you submit a request " +
+                    "re-using the same email for another harvest.") + "\n\n" +
+                    _("Thanks for supporting your community!") + "\n\n" +
+                    _("Yours") + ",\n" +
+                    "--\n" +
+                    "Saskatoon Harvest System")
```
Comment (2ynn) [inline saskatoon/harvest/signals.py:63]: I'm wondering if using a single string with the `format` method will not make translations easier. What do you think of trying with a syntax like this instead?
```
_("Hi  {} \n\n \
We are sorry but enough pickers have already been selected for the {}. \
You may still be contacted by the pick-leader if some of them end up cancelling. \ 
We will do our best to prioritize your participation next time you submit a request \
with the same email address. \n\n \
Thanks for supporting your community! \n\n \
Yours, \n -- \n \
Saskatoon Harvest System").format(picker_name, harvest_details)
```

Comment (victorphoenix3) [reply saskatoon/harvest/signals.py]: I tested and its working fine. Have updated the PR.

**Comment by victorphoenix3** [comment]
PR updated and rebased

**Review by tristanlatr — APPROVED** [review]
Looks good for me

**Comment by tristanlatr** [comment]
Thanks a lot for this @victorphoenix3 :)

**Comment by 2ynn** [comment]
> @2ynn are we good to merge this pull request ?
@tristanlatr yes let's go, do you have time to test all new automatic email functionalities on the prod server?

**Pull request merged**